d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -104,105 +107,160 @@ public WarmUpLoadBalancer(LoadBalancerWithFacilities balancer, WarmUpService ser
     _outstandingRequests = new ConcurrentLinkedDeque<>();
     _usedServices = new HashSet<>();
     _dualReadStateManager = dualReadStateManager;
+    _isIndis = isIndis;
+    _printName = String.format("%s WarmUp", _isIndis ? "xDS" : "ZK");
   }
 
   @Override
-  public void start(Callback<None> callback)
-  {
-    LOG.info("D2 WarmUp enabled");
-    _loadBalancer.start(new Callback<None>()
-    {
+  public void start(Callback<None> callback) {
+    LOG.info("{} enabled", _printName);
+    _loadBalancer.start(new Callback<None>() {
       @Override
-      public void onError(Throwable e)
-      {
+      public void onError(Throwable e) {
         callback.onError(e);
       }
 
       @Override
-      public void onSuccess(None result)
+      public void onSuccess(None result) {
+        _allStartTime = SystemClock.instance().currentTimeMillis();
+
+        Future<?> prepareTaskFuture = _executorService.submit(() -> prepareWarmUp());
+        try
+        {
+          prepareTaskFuture.get(_warmUpTimeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (TimeoutException e)
+        {
+          LOG.info("{} hit timeout: {}s. The WarmUp will continue in background", _printName, _warmUpTimeoutSeconds);
+          callback.onSuccess(None.none());
+        }
+        catch (Exception e)
```

Comment:
actually it's supposed to catch any kind of exception happened in the warmup (just like line 259), to make sure warmup won't fail startup since warmup is not required to complete successfully during startup.

Reply:
For that one usually catches `RuntimeException`, but it appears that that could just be an opinion.

Reply:
I understand. But to ensure no exception is raised up (including checked and runtime exceptions) from failures that we don't care, I think it's ok to catch the general exception/throwable (commonly seen in many mps).